### PR TITLE
Make CyRasterizer more configurable

### PR DESCRIPTION
**Pull request by trigeorgis**
Important changes:

The functionality to compile a new shader is done in the cython side of
the codebase. 

    class ShaderSource
      class VertexShader  < ShaderSource
      class FragmentShader < ShaderSource
      class GeometryShader < ShaderSource

There also exists a uniform manager which makes it possible to 
upload new uniform variables after a shader was compiled (GLUniform).

Whenever a user calls render_offscreen_rgb, we have to calculate the
normals of the mesh by having normalized array of normals, one per triangle
i.e., per triangle normals. This is done using numpy.cross.

The attributes in the shaders are hardcoded and are always in this order:

    layout(location = 0) in vec4 point;
    layout(location = 1) in vec2 tcoordIn;
    layout(location = 2) in vec3 linearMappingCoordIn;
    layout(location = 3) in vec3 normal;

### CODE REVIEW

File: cyrasterize/__init__.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
+import sys
 from cyrasterize.base import CyRasterizer
-
+from glrasterizer import FragmentShader, VertexShader
 
 from ._version import get_versions
 __version__ = get_versions()['version']
-del get_versions
+del get_versions
```
Comment (patricksnape) [inline cyrasterize/__init__.py]: New line missing

File: cyrasterize/cpp/glr.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #include "glrglfw.h"
 #include "stdint.h"
 #include <iostream>
+#include <python.h>
```
Comment (patricksnape) [inline cyrasterize/cpp/glr.h]: Why?

File: cyrasterize/shader.pyx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import logging as log
+
+from c_opengl_debug cimport *
+import sys
+
+__all__ = ['VertexShader', 'FragmentShader']
+
+cdef class ShaderSource:
+    # Most of this code comes from kivy
+    # https://github.com/kivy/kivy/blob/master/kivy/graphics/shader.pyx
```
Comment (patricksnape) [inline cyrasterize/shader.pyx]: Mention the Kivy licence? Possibly need a copy of their license?

File: cyrasterize/cpp/glr.h (hunk 2 above)
Comment (trigeorgis) [inline cyrasterize/cpp/glr.h]: Thanks it was unnecessary. Fixed in 54c3c53639b47426371c7e512a54916896ce1f36.

File: cyrasterize/shader.pyx (hunk 3 above)
Comment (trigeorgis) [inline cyrasterize/shader.pyx]: I had a look at Kivy source at the beginning as it was quite close to what we wanted by reimplemented most stuff from scratch as I found their code to be either a bit too complex or had a lot of features we didn't really need. I rephrased the comment in 2b038b6dc20b22072124db8a2afab7dbfca97f71.

**Comment by jabooth** [comment]
After my PR the code compiles successfully but then fails in testing with
```
Traceback (most recent call last):
  File "/Users/jab08/miniconda/conda-bld/test-tmp_dir/run_test.py", line 49, in <module>
    test_basic_random()
  File "/Users/jab08/miniconda/conda-bld/test-tmp_dir/run_test.py", line 32, in test_basic_random
    c = CyRasterizer(width=100, height=100)
  File "/Users/jab08/miniconda/envs/_test/lib/python2.7/site-packages/cyrasterize/base.py", line 104, in __init__
    self.uniforms = CyUniformBase(self._opengl)
  File "/Users/jab08/miniconda/envs/_test/lib/python2.7/site-packages/cyrasterize/base.py", line 24, in __init__
    uniforms = opengl.get_active_uniforms()
AttributeError: 'cyrasterize.glrasterizer.GLRasterizer' object has no attribute 'get_active_uniforms'
TESTS FAILED: cyrasterize-0.3-np11py27_0
```
Which seems odd - that's a def function in the superclass- George any ideas?

**Comment by patricksnape** [comment]
@trigeorgis Give @jabooth and I push access to your repo my man!

**Comment by jabooth** [comment]
The Menpobot failure is a problem with Jenkins - the build is now passing. LGTM! Thanks @trigeorgis!

**Pull request merged**